lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,471 @@
+                )
+            logger.info("TLS: Sent CA certificate to other relations")
+        else:
+            for relation in self.charm.model.relations.get(SEND_CA_CERT_RELATION_NAME, []):
+                self.certificate_transfer.remove_certificate(relation.id)
+            logger.info("TLS: Removed CA cert from relations")
+
+    def _generate_self_signed_certs(self, ingress_address: str) -> None:
+        """Recreates a unit certificate from the Vault CA certificate, then saves it.
+
+        Args:
+            ingress_address: The subject of the unit certificate.
+        """
+        self._remove_all_certs_from_workload()
+
+        if not (private_key := self._pull_tls_file_from_workload(File.KEY)):
+            private_key = generate_private_key().decode()
+            self._push_tls_file_to_workload(File.KEY, private_key)
+
+        ca_private_key, ca_certificate = self._get_ca_certificate_secret_in_peer_relation()
+        self._push_tls_file_to_workload(File.CA, ca_certificate)
+        sans_ip = [ingress_address]
+        certificate = generate_vault_unit_certificate(
+            subject=ingress_address,
+            sans_ip=sans_ip,
+            sans_dns=[socket.getfqdn()],
+            ca_certificate=ca_certificate.encode(),
+            ca_private_key=ca_private_key.encode(),
+            unit_private_key=private_key.encode(),
+        )
+        self._push_tls_file_to_workload(File.CERT, certificate)
+
+    def _new_certificate_request_pending(self) -> bool:
+        """Check if we have a pending CSR by comparing the requirer and provider databags."""
+        # TODO: This will be a library function in the future
+        created_csr = self.tls_access._requirer_csrs
+        return (
+            False
+            if any(
+                [
+                    self.tls_access._find_certificate_in_relation_data(
+                        entry["certificate_signing_request"]  # type: ignore [arg-type]
+                    )
+                    for entry in created_csr
+                ]
+            )
+            else True
+        )
+
+    def _should_request_new_certificate(self) -> bool:
+        """Determines if we should request a new certificate from the tls relation."""
+        if not self.charm.model.relations.get(TLS_CERTIFICATE_ACCESS_RELATION_NAME):
+            return False
+        csr_exists = (existing_csr := self._pull_tls_file_from_workload(File.CSR))
+        if not csr_exists:
+            return True
+        csr_is_for_current_certificate = csr_matches_certificate(
+            existing_csr, self._pull_tls_file_from_workload(File.CERT)
+        )
+        certificate_request_pending = self._new_certificate_request_pending()
+        new_certificate_available = self.tls_access._find_certificate_in_relation_data(
+            existing_csr
+        )
+        certificate_expiring = False
+        # TODO: we probably should prefer either a CRL or a method in the library for this
```

Comment:
CRLs would be a good idea to implement in general for TLS, however it requires a stateful server that can handle it, and it would need to be designed for the whole ecosystem. Maybe we should have a design task to spec this out instead of leaving it in a TODO?

Reply:
I agree. I can erase it, since I don't think anyone will have an expiring certificate in production using vault anytime soon. I can also use the `certificate expiring` event to remove the certificate, but my understanding from conversations with @gruyaume is that we would like to avoid taking logic out of the configure function.

Reply:
I just added a new function to the library that achieves this for now. CRL would be a bit overkill just for it to have a single use here in Vault in my opinion. Once the library is updated I will use that here and remove the TODO.

Reply:
> I just added a new function to the library that achieves this for now

Can you link to this? I can't follow the conversation very well.

Reply:
The library is now updated, and the changes are merged here. There's no need for a CRL, and so there is no longer any reason to discuss this right now in my opinion.